akka-http-core/src/main/scala/akka/http/impl/engine/parsing/ParserOutput.scala

```diff
@@ -64,7 +65,11 @@ private[http] object ParserOutput {
   sealed abstract class EntityCreator[-A <: ParserOutput, +B >: HttpEntity.Strict <: HttpEntity] extends (Source[A, Unit] ⇒ B)
 
   final case class StrictEntityCreator(entity: HttpEntity.Strict) extends EntityCreator[ParserOutput, HttpEntity.Strict] {
-    def apply(parts: Source[ParserOutput, Unit]) = entity
+    def apply(parts: Source[ParserOutput, Unit]) = {
+      // We might need to drain stray empty tail streams which will be read by no one.
+      parts.runWith(Sink.ignore)(GraphInterpreter.currentInterpreter.subFusingMaterializer)
```

Comment:
Unfortunately this was a bug, even in the non-fused version, it just needed a much rare race, mostly avoided by the buffers+prefetch. Currently this is deterministic due to lack of buffers and prefetch. This can be probably hacked to be faster, but I think there is a superior solution that you already hinted at in another PR.

I think the proper solution is to put the substream streaming right into the parser, avoiding the `splitWhen` dance altogether. The benefits would be immense:
 - not creating accidental empty substreams just to drop them (and materialize them), since the parser knows when to emit streams in the first place
 - getting rid of four stages: `splitWhen`, `prefixAndTail`, `collect`, and `concatSubstreams`. This means less memory, and less execution cost.
 - getting rid of two async callbacks: Currently one in `splitWhen`, `prefixAndTail` and `concatSubstreams`. These would be replaced with only one in the parser
 - getting rid of the materialization in `concatSubstreams`. This one would be huge alone.
 - strict entities would not need any substreams at all. This means that the cost of substreams would be only paid for entities that need it, strict entities would not involve any substreams at all (and these are very common for ordinary requests, so this is a hot path).
 - it would be possible to plant an `AsyncCallback` in non-strict entities that allow cancellation of the entity stream without materialization with `runWith(Sink.cancel)`. This can be very nice for automatic substream cleanup

One cost of this change would be that it complicates the parser. Creating a stream-of-streams `GraphStage` is not a simple task, as it is shown by `SplitWhen`. On the other hand this change would simplify the overall HTTP architecture quite a bit.

The other objection can be philosophical: creating custom stream-of-stream stages for each use case is something to be avoided in general, and not something we should advertise or practice, but the current solution here is both costly and also not "pure". I think until we find some more general, philosophically desirable solution, the performance benefits of this change waaaay outweigh the costs of purity here.

/cc @sirthias @rkuhn @ktoso